### ISSUE TITLE
connectionTimeout vs. slow first connection attempt

### ISSUE THREAD

**Issue by patric-r** [brettwooldridge/HikariCP#614]
Please consider this as a feature request.

We use a tight `connectionTimeout` (5s) in some environments to quickly identify issues with database connection attempts. During regular application runs, this works perfectly and as expected.

From time to time, we're facing a timeout exception during startup of some of our applications.
During analysis we found out that there are many things which will slow down the first connection attempt within the jdbc driver, e.g.:
- classloading (file system access)
- jar verification / integrity check
- crypto subsystem initialization (`SecretKeyFactory` ...)
- byte code transforming (jacoco ...)

If more than one HikariCP-Pool is used / initialized at the same time, things get even worse with some jdbc drivers because thread synchronization blocks HikariCP's connection adder threads, as a thread dumps shows:

```
     [java] "HikariPool-3 connection adder" Id=15 BLOCKED on sun.security.jca.ProviderConfig@70be0a2b owned by "HikariPool-1 connection adder" Id=16 
     [java]     at sun.security.jca.ProviderConfig.getProvider(ProviderConfig.java:168)
     [java]     -  blocked on sun.security.jca.ProviderConfig@70be0a2b
     [java]     at sun.security.jca.ProviderList.getProvider(ProviderList.java:233)
     [java]     at sun.security.jca.ProviderList$ServiceList.tryGet(ProviderList.java:434)
     [java]     at sun.security.jca.ProviderList$ServiceList.access$200(ProviderList.java:376)
     [java]     at sun.security.jca.ProviderList$ServiceList$1.hasNext(ProviderList.java:486)
     [java]     at javax.crypto.SecretKeyFactory.nextSpi(SecretKeyFactory.java:293)
     [java]     -  locked java.lang.Object@4c15e7fd 
     [java]     at javax.crypto.SecretKeyFactory.<init>(SecretKeyFactory.java:121)
     [java]     at javax.crypto.SecretKeyFactory.getInstance(SecretKeyFactory.java:160)
     [java]     at oracle.jdbc.driver.T4C8TTIdty.<init>(T4C8TTIdty.java:517)
     [java]     at oracle.jdbc.driver.T4CConnection.connect(T4CConnection.java:1603)
     [java]     at oracle.jdbc.driver.T4CConnection.logon() line: 496   
     [java]     at oracle.jdbc.driver.T4CConnection(oracle.jdbc.driver.PhysicalConnection).connect() line: 666  
     [java]     at oracle.jdbc.driver.T4CDriverExtension.getConnection(java.lang.String, java.util.Properties) line: 32 
     [java]     at oracle.jdbc.OracleDriver(oracle.jdbc.driver.OracleDriver).connect(java.lang.String, java.util.Properties) line: 566  
     [java]     at com.zaxxer.hikari.util.DriverDataSource.getConnection() line: 95 [HikariCP-2.4.5.jar:?]
[...]
```

Of course, HikariCP cannot be blamed for this.
However, because of all those reasons, a tight `connectionTimeout` might be a problem for the first connection attempt.

A separate connection timeout for the first attempt (or something similar) would really help us here.

**Comment by nitincchauhan** [comment]
@patric-r 
check initializationFailFast, with true (default), Creating first connection will not use connectionTimeout for time out.

**Comment by patric-r** [comment]
@nitincchauhan 
You're right. However, the one thing what I don't like with `initializationFailFast` is the fact that this "initialization test connection" will now be put into the pool but closed right after its acquisition.
Especially for very short pool life times which might happen in several circumstances (e.g. unit testing), it adds a significant overhead to the process.

**Comment by brettwooldridge** [comment]
It used to be added to the pool.  We'll take a look at the commit history and see if we can return to that behavior (or if it addressed another issue).

**Comment by patric-r** [comment]
Thank you, Brett.
